File: python_analyzer_service/api_client.py
# python_analyzer_service/api_client.py
import requests
import logging
import os
import json
from typing import Dict, Any

logger = logging.getLogger(__name__)

# Get API Gateway endpoint from environment variable or use default
# Default assumes docker-compose network alias 'api_gateway' on port 8000
API_GATEWAY_ENDPOINT = os.getenv('API_GATEWAY_ENDPOINT', 'http://api_gateway:8043') # Updated default port
INGEST_URL = f"{API_GATEWAY_ENDPOINT}/ingest/analysis_data"
REQUEST_TIMEOUT = 30 # Timeout in seconds

def send_analysis_data(data: Dict[str, Any]):
    """
    Sends the formatted analysis data (nodes and relationships) to the API Gateway ingestion endpoint.

    Args:
        data: A dictionary containing 'filePath', 'nodes', and 'relationships'.
    """
    file_path = data.get("filePath", "unknown file")
    logger.info(f"Attempting to send analysis data for {file_path} to {INGEST_URL}")

    try:
        # Construct the payload expected by the API Gateway
        payload_to_send = {
            "nodes": data.get("nodes", []),
            "relationships": data.get("relationships", [])
        }
        response = requests.post(
            INGEST_URL,
            json=payload_to_send, # Send the correctly structured payload
            timeout=REQUEST_TIMEOUT,
            headers={'Content-Type': 'application/json'}
        )
        response.raise_for_status() # Raises HTTPError for bad responses (4xx or 5xx)
        logger.info(f"Successfully sent analysis data for {file_path}. Status code: {response.status_code}")
        return True

    except requests.exceptions.Timeout:
        logger.error(f"Timeout error sending analysis data for {file_path} to {INGEST_URL} after {REQUEST_TIMEOUT} seconds.")
        return False
    except requests.exceptions.ConnectionError as conn_err:
        logger.error(f"Connection error sending analysis data for {file_path} to {INGEST_URL}: {conn_err}")
        return False
    except requests.exceptions.HTTPError as http_err:
        logger.error(f"HTTP error sending analysis data for {file_path}: {http_err.response.status_code} - {http_err.response.text}")
        return False
    except requests.exceptions.RequestException as req_err:
        logger.exception(f"An unexpected error occurred sending analysis data for {file_path}: {req_err}")
        return False
    except Exception as e:
        # Catch any other unexpected errors during the process
        logger.exception(f"A non-request error occurred during API submission for {file_path}: {e}")
        return False

File: javascript_analyzer_service/api_client.js
// javascript_analyzer_service/api_client.js
// Handles sending analysis data to the ingestion API gateway

const axios = require('axios');

// TODO: Make this configurable via environment variables
const API_GATEWAY_ENDPOINT = 'http://api_gateway:8000/ingest/analysis_data';

/**
 * Sends the analyzed node and relationship data to the ingestion API.
 * @param {object} analysisData - The data object containing 'nodes' and 'relationships' arrays.
 * @param {string} filePath - The path of the file analyzed (for logging purposes).
 * @returns {Promise<boolean>} - True if successful, false otherwise.
 */
async function sendAnalysisData(analysisData, filePath) {
    if (!analysisData || !analysisData.nodes || !analysisData.relationships) {
        console.error(`[API Client] Invalid analysis data provided for ${filePath}. Aborting send.`);
        return false;
    }

    console.log(`[API Client] Sending analysis data for ${filePath} to ${API_GATEWAY_ENDPOINT}...`);
    console.log(`[API Client] Data size: ${analysisData.nodes.length} nodes, ${analysisData.relationships.length} relationships.`);

    try {
        const response = await axios.post(API_GATEWAY_ENDPOINT, analysisData, {
            headers: { 'Content-Type': 'application/json' },
            // Consider adding a timeout
            // timeout: 10000, // e.g., 10 seconds
        });

        if (response.status >= 200 && response.status < 300) {
            console.log(`[API Client] Successfully sent analysis data for ${filePath}. Status: ${response.status}`);
            return true;
        } else {
            console.error(`[API Client] API Gateway returned non-success status for ${filePath}: ${response.status} ${response.statusText}`);
            console.error(`[API Client] Response data:`, response.data);
            return false;
        }
    } catch (error) {
        console.error(`[API Client] Error sending analysis data for ${filePath}:`, error.message);
        if (error.response) {
            // The request was made and the server responded with a status code
            // that falls out of the range of 2xx
            console.error(`[API Client] Error Response Status: ${error.response.status}`);
            console.error(`[API Client] Error Response Data:`, error.response.data);
        } else if (error.request) {
            // The request was made but no response was received
            console.error('[API Client] No response received from API Gateway. Check network connectivity and endpoint availability.');
            console.error('[API Client] Error Request:', error.request);
        } else {
            // Something happened in setting up the request that triggered an Error
            console.error('[API Client] Error setting up request:', error.message);
        }
        return false;
    }
}

module.exports = { sendAnalysisData };